// @phala/fn/dist/commands/run.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const node_fs_1 = require("node:fs");
const core_1 = require("@oclif/core");
const utils_1 = require("../lib/utils");
const runQuickJs_1 = require("../lib/runQuickJs");
class Run extends core_1.Command {
    async run() {
        const { flags: { scriptArgs = [] }, args: { script }, } = await this.parse(Run);
        const scriptPath = (0, utils_1.resolveToAbsolutePath)(script);
        const js = (0, node_fs_1.readFileSync)(scriptPath, 'utf8');
        const output = await (0, runQuickJs_1.runQuickJs)(js, scriptArgs, { silent: this.jsonEnabled() });
        this.log(JSON.stringify({ output }));
        return {
            output,
        };
    }
}
Run.description = 'Run JS in QuickJS runtime';
Run.enableJsonFlag = true;
Run.flags = {
    scriptArgs: core_1.Flags.string({
        char: 'a',
        description: 'Script arguments',
        multiple: true,
    }),
};
Run.args = {
    script: core_1.Args.string({
        description: 'The location of the JS file',
        required: true,
    }),
};
exports.default = Run;
