// @phala/fn/dist/lib/BaseCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const ora_1 = tslib_1.__importDefault(require("ora"));
class BaseCommand extends core_1.Command {
    constructor() {
        super(...arguments);
        this.spinner = (0, ora_1.default)();
    }
    get action() {
        return {
            start: (message) => {
                this.spinner.stop();
                this.spinner.start(message);
            },
            stop: () => this.spinner.stop(),
            succeed: (message) => this.spinner.stopAndPersist({
                symbol: chalk_1.default.green('✓'),
                text: message,
            }),
            fail: (message) => {
                this.spinner.stopAndPersist({
                    symbol: chalk_1.default.red('✗'),
                    text: message,
                });
            },
            warn: (message) => {
                this.spinner.stopAndPersist({
                    symbol: chalk_1.default.yellow('!'),
                    text: message,
                });
            },
        };
    }
}
exports.default = BaseCommand;
