// quickjs-emscripten/dist/variants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RELEASE_ASYNC = exports.DEBUG_ASYNC = exports.RELEASE_SYNC = exports.DEBUG_SYNC = exports.memoizePromiseFactory = exports.newQuickJSAsyncWASMModule = exports.newQuickJSWASMModule = void 0;
const esmHelpers_1 = require("./esmHelpers");
/**
 * Create a new, completely isolated WebAssembly module containing the QuickJS library.
 * See the documentation on [[QuickJSWASMModule]].
 *
 * Note that there is a hard limit on the number of WebAssembly modules in older
 * versions of v8:
 * https://bugs.chromium.org/p/v8/issues/detail?id=12076
 */
async function newQuickJSWASMModule(
/**
 * Optionally, pass a {@link SyncBuildVariant} to construct a different WebAssembly module.
 */
variant = exports.RELEASE_SYNC) {
    const [wasmModuleLoader, QuickJSFFI, { QuickJSWASMModule }] = await Promise.all([
        variant.importModuleLoader(),
        variant.importFFI(),
        import("./module.js").then(esmHelpers_1.unwrapTypescript),
    ]);
    const wasmModule = await wasmModuleLoader();
    wasmModule.type = "sync";
    const ffi = new QuickJSFFI(wasmModule);
    return new QuickJSWASMModule(wasmModule, ffi);
}
exports.newQuickJSWASMModule = newQuickJSWASMModule;
/**
 * Create a new, completely isolated WebAssembly module containing a version of the QuickJS library
 * compiled with Emscripten's [ASYNCIFY](https://emscripten.org/docs/porting/asyncify.html) transform.
 *
 * This version of the library offers features that enable synchronous code
 * inside the VM to interact with asynchronous code in the host environment.
 * See the documentation on [[QuickJSAsyncWASMModule]], [[QuickJSAsyncRuntime]],
 * and [[QuickJSAsyncContext]].
 *
 * Note that there is a hard limit on the number of WebAssembly modules in older
 * versions of v8:
 * https://bugs.chromium.org/p/v8/issues/detail?id=12076
 */
async function newQuickJSAsyncWASMModule(
/**
 * Optionally, pass a {@link AsyncBuildVariant} to construct a different WebAssembly module.
 */
variant = exports.RELEASE_ASYNC) {
    const [wasmModuleLoader, QuickJSAsyncFFI, { QuickJSAsyncWASMModule }] = await Promise.all([
        variant.importModuleLoader(),
        variant.importFFI(),
        import("./module-asyncify.js").then(esmHelpers_1.unwrapTypescript),
    ]);
    const wasmModule = await wasmModuleLoader();
    wasmModule.type = "async";
    const ffi = new QuickJSAsyncFFI(wasmModule);
    return new QuickJSAsyncWASMModule(wasmModule, ffi);
}
exports.newQuickJSAsyncWASMModule = newQuickJSAsyncWASMModule;
/**
 * Helper intended to memoize the creation of a WebAssembly module.
 * ```typescript
 * const getDebugModule = memoizePromiseFactory(() => newQuickJSWASMModule(DEBUG_SYNC))
 * ```
 */
function memoizePromiseFactory(fn) {
    let promise;
    return () => {
        return (promise ??= fn());
    };
}
exports.memoizePromiseFactory = memoizePromiseFactory;
/**
 * This build variant is compiled with `-fsanitize=leak`. It instruments all
 * memory allocations and when combined with sourcemaps, can present stack trace
 * locations where memory leaks occur.
 *
 * See [[TestQuickJSWASMModule]] which provides access to the leak sanitizer via
 * {@link TestQuickJSWASMModule.assertNoMemoryAllocated}.
 *
 * The downside is that it's 100-1000x slower than the other variants.
 * Suggested use case: automated testing, regression testing, and interactive
 * debugging.
 */
exports.DEBUG_SYNC = {
    type: "sync",
    async importFFI() {
        const mod = await import("./generated/ffi.WASM_DEBUG_SYNC.js");
        return (0, esmHelpers_1.unwrapTypescript)(mod).QuickJSFFI;
    },
    async importModuleLoader() {
        const mod = await import("./generated/emscripten-module.WASM_DEBUG_SYNC.js");
        return (0, esmHelpers_1.unwrapJavascript)(mod).default;
    },
};
/**
 * This is the default (synchronous) build variant.
 * {@link getQuickJS} returns a memoized instance of this build variant.
 */
exports.RELEASE_SYNC = {
    type: "sync",
    async importFFI() {
        const mod = await import("./generated/ffi.WASM_RELEASE_SYNC.js");
        return (0, esmHelpers_1.unwrapTypescript)(mod).QuickJSFFI;
    },
    async importModuleLoader() {
        const mod = await import("./generated/emscripten-module.WASM_RELEASE_SYNC.js");
        return (0, esmHelpers_1.unwrapJavascript)(mod).default;
    },
};
/**
 * The async debug build variant may or may not have the sanitizer enabled.
 * It does print a lot of debug logs.
 *
 * Suggested use case: interactive debugging only.
 */
exports.DEBUG_ASYNC = {
    type: "async",
    async importFFI() {
        const mod = await import("./generated/ffi.WASM_DEBUG_ASYNCIFY.js");
        return (0, esmHelpers_1.unwrapTypescript)(mod).QuickJSAsyncFFI;
    },
    async importModuleLoader() {
        const mod = await import("./generated/emscripten-module.WASM_DEBUG_ASYNCIFY.js");
        return (0, esmHelpers_1.unwrapJavascript)(mod).default;
    },
};
/**
 * This is the default asyncified build variant.
 */
exports.RELEASE_ASYNC = {
    type: "async",
    async importFFI() {
        const mod = await import("./generated/ffi.WASM_RELEASE_ASYNCIFY.js");
        return (0, esmHelpers_1.unwrapTypescript)(mod).QuickJSAsyncFFI;
    },
    async importModuleLoader() {
        const mod = await import("./generated/emscripten-module.WASM_RELEASE_ASYNCIFY.js");
        return (0, esmHelpers_1.unwrapJavascript)(mod).default;
    },
};
